app/lib/pdf/finance/invoice_report_data.rb

```diff
@@ -51,17 +52,15 @@ def plan
     @invoice.buyer_account.bought_plan.name
   end
 
-  def logo
-    @logo_stream ||= open(@invoice.provider_account.profile.logo.url(:invoice))
-  rescue StandardError => e
-    # TODO: - uncomment complete exception!
-    # rescue OpenURI::HTTPError => e
-    Rails.logger.error "Failed to retrieve logo from: #{e.message}"
-    nil
-  end
-
-  def logo?
-    @invoice.provider_account.profile.logo.file? && logo
+  def with_logo
+    file = logo_file
+    yield file if block_given?
+  rescue StandardError => exception
+    Rails.logger.error "Failed to retrieve logo: #{exception.message}"
+    yield nil if block_given?
+  ensure
+    file ||= nil # initialize the variable if it is not initialized yet
+    file&.close
```

Comment:
Regarding the "cognitive complexity" issue, would this work?
```ruby
def with_logo
  file = begin
    logo_file
  rescue StandardError => exception
    Rails.logger.error "Failed to retrieve logo: #{exception.message}"
  end

  yield file if block_given?
  file.close if file
end
```
From what I could gather, this `logo_file` method is the one that could raise an `StandardError`. If that's the case, we can isolate the rescues to it's assignment.

Reply:
Thanks @thalesmiguel I can try that!

`file.close if file`though gives:
```
RuboCop: Use safe navigation (`&.`) instead of checking if an object exists before calling the method. [Style/SafeNavigation]
```
and
```
Local variable 'file' may be uninitialized 
```
That's why this ugly stuff.

But maybe the 2nd one is just the IDE warning, and CodeClimate will be happy. I'll try.

Reply:
Actually, do we need this `yield nil if block_given?`?
If not, this could also be
```ruby
def with_logo
  yield logo_file if block_given?
  logo_file.close
rescue StandardError => exception
  Rails.logger.error "Failed to retrieve logo: #{exception.message}"
end
```
From the implementation of `logo_file`, I think it will either open a file or raise an error. So, if there are no errors, we can assume the file is open and we can close it. Does that make any sense? 😅

Reply:
> ```
> Local variable 'file' may be uninitialized 
> ```

I get Rubocop complains about the `if file` 😬 Let's play safe with `&` then... About this safe navigation, I usually try my best to not use it, as it's a "`nil` check", which could hide some odd behaviour. Just a rule of thumb for me (that I break from time to time).

Now this `Local variable 'file' may be uninitialized` is odd, as `file` will always be initialized, even if there's an error on the block that initializes it... Maybe Rubocop isn't able to see it.
I'm talking about something like this:
```ruby
a_new_variable = begin
  raise StandardError
rescue StandardError
end #=> nil

a_new_variable #=> nil (no kaboom)
```

Reply:
Oh, yes, with this change (var = begin) indeed it will always be initialized. But we do need to keep `nil`, we want to know when the logo is not available, then we won't print it on the invoice.

And also, we actually will need to keep file closing in `ensure`, because we can't trust that the code in the block will do something nasty, see https://railsware.com/blog/yield-gotcha-in-ruby-blocks/

Reply:
Ah yes. That was the question regarding that `yield`. If something shady happens there and the execution stops (with a `raise`, for example), the file would not be closed :v:

Reply:
It is always initialized even if you have:
```ruby
if something
  test = :foo
else
  whatever else
end

# test is either `:foo` or `nil` here 
```

Reply:
Oh @akostadinov I didn't know that!
The case was actually different (no if), but anyway - I think I am on the right track :wink:

Reply:
Same goes for
```
def methodx
  raise
  test = :foo
ensure
  # test is `nil` here
end
```

But I don't know if this style is discouraged for some reason.